pkg/controllers/user/authz/podsecuritypolicy/serviceaccount.go

```diff
@@ -27,213 +26,265 @@ func RegisterServiceAccount(context *config.UserContext) {
-	namespaceLister v13.NamespaceLister
-	projectLister   v3.ProjectLister
+	clusterName       string
+	clusterLister     v3.ClusterLister
+	clusters          v3.ClusterInterface
+	pspts             v3.PodSecurityPolicyTemplateInterface
+	psptLister        v3.PodSecurityPolicyTemplateLister
+	templateLister    v3.PodSecurityPolicyTemplateLister
+	policyLister      v1beta1.PodSecurityPolicyLister
+	roleBindingLister v12.RoleBindingLister
+	roleBindings      v12.RoleBindingInterface
+	policies          v1beta1.PodSecurityPolicyInterface
+	roleLister        v12.ClusterRoleLister
+	roles             v12.ClusterRoleInterface
+	namespaceLister   v13.NamespaceLister
+	projectLister     v3.ProjectLister
+	psptpbLister      v3.PodSecurityPolicyTemplateProjectBindingLister
 }
 
-func (m *serviceAccountManager) sync(key string, obj *v1.ServiceAccount) error {
-	if obj == nil {
-		logrus.Debugf("no service account provided, exiting")
-		return nil
-	}
+func (m *serviceAccountManager) Create(obj *v1.ServiceAccount) (*v1.ServiceAccount, error) {
+	return m.sync(obj)
+}
 
-	namespaceName := obj.Namespace
+func (m *serviceAccountManager) Updated(obj *v1.ServiceAccount) (*v1.ServiceAccount, error) {
+	return m.sync(obj)
+}
 
-	// get PSPT
-	// if no PSPT then get default
-	// if no default then exit
-	namespace, err := m.namespaceLister.Get("", namespaceName)
+func (m *serviceAccountManager) sync(obj *v1.ServiceAccount) (*v1.ServiceAccount, error) {
+	namespace, err := m.namespaceLister.Get("", obj.Namespace)
 	if err != nil {
-		return fmt.Errorf("error getting namespaces: %v", err)
+		return nil, fmt.Errorf("error getting projects: %v", err)
 	}
 
-	annotation := namespace.Annotations[projectIDAnnotation]
-
-	if annotation == "" {
-		// no project is associated with namespace so don't do anything
-		logrus.Debugf("no project is associated with namespace %v", namespaceName)
-		return nil
+	if namespace.Annotations[projectIDAnnotation] == "" {
+		return obj, nil
 	}
 
-	split := strings.SplitN(annotation, ":", 2)
-
-	if len(split) != 2 {
-		return fmt.Errorf("could not parse handler key: %v got len %v", annotation, len(split))
-	}
-
-	clusterName, projectID := split[0], split[1]
-
-	podSecurityPolicyTemplateID, err := getPodSecurityPolicyTemplateID(m.projectLister, m.clusterLister, projectID,
-		clusterName)
+	unfilteredPSPTPBS, err := m.psptpbLister.List("", labels.Everything())
 	if err != nil {
-		return err
+		return nil, fmt.Errorf("error getting psptpbs: %v", err)
 	}
 
-	if podSecurityPolicyTemplateID == "" {
-		// Do nothing
-		logrus.Debugf("no matching pod security policy template")
-		return nil
+	var psptpbs []*v3.PodSecurityPolicyTemplateProjectBinding
+	for _, psptpb := range unfilteredPSPTPBS {
+		if psptpb.DeletionTimestamp == nil {
+			psptpbs = append(psptpbs, psptpb)
+		}
 	}
 
-	policies, err := m.policyLister.List("", labels.Everything())
+	roleBindings, err := m.roleBindingLister.List(obj.Namespace, labels.Everything())
```

Comment:
scares me to get *all* bindings in the namespace here. you should _only_ get bindings that have an owner reference of the current service account. You would need to have an indexer that returns bindings based on the owner reference UID